### PR TITLE
Dev kaijie

### PR DESCRIPTION
**Pull request by KaijieMo1**
- add network body part identification
- update patch method  in the pipeline

### CODE REVIEW

File: plot/plot_figure.py [hunk 1]
```diff
@@ -353,68 +373,238 @@ def plot_area_ratio(config, list_images_series, slice_dim=2, merge_channel_plot=
...
     if rotate:
         image = np.rot90(image, k=1, axes=(1, 0))
         heatmap = np.rot90(heatmap, k=1, axes=(1, 0))
-        aspect = 1.0 / aspect
 
-    # Discrete color scheme
-    cMap = ListedColormap(['midnightblue', 'darkslateblue', 'darkcyan', 'olive', 'goldenrod'])
 
+    # Discrete color scheme
+    cMap = ListedColormap([[0.1,0.2,0.3], [0.1,0.3,0.4],[0.3,0.3,0.3], [0.6,0.4,0.4],[0.4,0.5,0.8], [0.7,0.2,0.5],])
```
Comment (thomaskuestner) [inline plot/plot_figure.py]: load from config

File: plot/plot_figure.py [hunk 2]
```diff
@@ -353,68 +373,238 @@ def plot_area_ratio(config, list_images_series, slice_dim=2, merge_channel_plot=
...
     fig, ax = plt.subplots()
     image = ax.pcolormesh(image)
     heatmap = ax.pcolor(heatmap, alpha=alpha, cmap=cMap)
-    ax.set_aspect(aspect=aspect)
 
     cbar = plt.colorbar(heatmap)
+
     landmark_dict = {0: 'Wrists', 1: 'Shoulders', 2: 'Liver_dome', 3: 'Hips', 4: 'Heels', 5: 'Below'}
-    cbar.ax.set_ylabel('Class')
+    cbar.ax.set_ylabel('Class',fontsize=20)
```
Comment (thomaskuestner) [inline plot/plot_figure.py]: load fontsize from config

File: models/loss_function.py [hunk 3]
```diff
@@ -4,80 +4,53 @@
...
-def dice_loss_v2(y_true, y_pred,config):
     smooth = 1E-16
+    # smooth = K.epsilon()
     sum_loss, weight_sum = 0, 0
     for class_index in range(config['channel_label_num']):
         y_t = y_true[..., class_index]
         y_p = y_pred[..., class_index]
-        intersection = K.sum(K.abs(y_t * y_p), axis=-1)
-        loss = 1 - (2. * intersection + smooth) / (K.sum(K.abs(y_t)) + K.sum(K.abs(y_p)) + smooth)
+        intersection = K.sum(K.abs(y_t * y_p))
```
Comment (thomaskuestner) [inline models/loss_function.py:44]: can't accept redefinition of dice loss -> keep compatibility. Is axis=-1 needed?

File: models/loss_function.py [hunk 4]
```diff
@@ -4,80 +4,53 @@
...
-    for class_index in range(config['channel_label_num']):
-        y_t=y_true[...,class_index]
-        y_p=y_pred[...,class_index]
-        intersection = K.sum(K.abs(y_t * y_p), axis=-1)
-        loss =1 - (2. * intersection + smooth) / (K.sum(K.square(y_t), -1) + K.sum(K.square(y_p), -1) + smooth)
-        sum_loss+=loss*config['loss_channel_weight'][class_index]
-        weight_sum+=config['loss_channel_weight'][class_index]
-    return sum_loss/(weight_sum+smooth)
-
-def dice_coefficient_loss(y_true, y_pred,config, smooth=K.epsilon(), axis=None):
```
Comment (thomaskuestner) [inline models/loss_function.py:47]: dice_coefficient_loss is required for organ segmentation

File: models/loss_function.py [hunk 5]
```diff
@@ -4,80 +4,53 @@
...
-    """ Dice coefficient along specific axis (same as  1+dice_loss() if axis=None)
-            y_true: true targets tensor.
-            y_pred: predictions tensor.
-            smooth: smoothing parameter to avoid division by zero
-            axis: along which to calculate Dice
-    """
-    intersection = K.sum(K.abs(y_true * y_pred), axis=axis)
-    return -(2. * intersection + smooth) / (K.sum(K.abs(y_true), axis=axis) + K.sum(K.abs(y_pred), axis=axis) + smooth)
 
-def dice_loss_v2(y_true, y_pred,config):
```
Comment (thomaskuestner) [inline models/loss_function.py:57]: critical redefinition of interfaces

File: models/ModelSet.py [hunk 6]
```diff
@@ -147,167 +149,92 @@ def model_MRGE_1(self, config):
...
-        if config['feed_pos']:
-            return create_and_compile_model([inputs, in_pos], out, config)
-        else:
-            return create_and_compile_model(inputs, out, config)
-
+    def model_position_densenet(self, config):
+        conv_param = config['convolution_parameter']
 
-    def model_U_net_old(self, config, depth=None):
+        def denseBlock(dense_block_len, filters):
```
Comment (thomaskuestner) [inline models/ModelSet.py:173]: relevant blocks should be defined in blocks.py and not ModelSet.py -> move

File: models/ModelSet.py [hunk 7]
```diff
@@ -147,167 +149,92 @@ def model_MRGE_1(self, config):
...
-            return create_and_compile_model([inputs, in_pos], out, config)
-        else:
-            return create_and_compile_model(inputs, out, config)
+        # filters = [config['filters'] * 2 ** i for i in range(5)]
+        # filters = [config['filters']*2, config['filters']*2,config['filters']*2,config['filters']*4 ]
+        filters = [8, 8, 8, 12]
+        f2=[32,64,128,128]
 
-    # UNet with (noised) positional input (feed_pos) and in-block skips
-    def model_U_net(self, config):
```
Comment (thomaskuestner) [inline models/ModelSet.py:250]: can't accept deletion of unet model

File: models/ModelSet.py [hunk 8]
```diff
@@ -147,167 +149,92 @@ def model_MRGE_1(self, config):
...
-
-        if config['feed_pos']:
-            return create_and_compile_model([inputs, in_pos], out, config)
-        else:
-            return create_and_compile_model(inputs, out, config)
-
+    def model_position_densenet(self, config):
+        conv_param = config['convolution_parameter']
 
-    def model_U_net_old(self, config, depth=None):
```
Comment (thomaskuestner) [inline models/ModelSet.py:218]: can't accept deletion

File: models/ModelSet.py [hunk 9]
```diff
@@ -361,133 +289,36 @@ def dense(x, f, rates, conv_param):
...
-        x = block(filters[0], 4, 2, order=['dc', 'b', 'r'], order_param=[conv_param, None, None])(x)
-        out = block(config['channel_label_num'], 4, 1, order=['c', 'b', 's'], order_param=[conv_param, None, None])(x)
-
-        if config['feed_pos']:
-            return create_and_compile_model([inputs, in_pos], out, config)
-        else:
-            return create_and_compile_model(inputs, out, config)
-
-    def model_U_net_2double_decoder(self, config):
-        """2 cascade U Net double decoder"""
```
Comment (thomaskuestner) [inline models/ModelSet.py:487]: were the UNet models merged in simpler implementation? If so interfaces of these old models should be kept for backward comp. and pointing to new implementations

File: train.py [hunk 10]
```diff
@@ -8,8 +8,6 @@
 from models.load_model import load_model_file
 import pickle
 import datetime
-import os
```
Comment (thomaskuestner) [inline train.py:12]: keep import os

File: train.py [hunk 11]
```diff
@@ -208,20 +230,18 @@ def k_fold_train_process(config, model, k_fold, paths, dataset, cp_callback, ini
         if not config['k_fold_merge_model']:
 
             # train all k-fold on one model
-            model, history_curr = train_process(config, model, paths_train_img, paths_train_label, paths_val_img,
```
Comment (thomaskuestner) [inline train.py:215]: redefinition can't be accepted. history needs to be saved as list over k-fold runs

**Review by thomaskuestner — CHANGES_REQUESTED** [review]
Major changes in loss and model can't be accepted. Backward compatibility not ensured.

File: models/loss_function.py (hunk 3 above)
Comment (KaijieMo1) [reply models/loss_function.py:44]: The axis=-1 should be deleted.

File: models/ModelSet.py (hunk 9 above)
Comment (KaijieMo1) [reply models/ModelSet.py:487]: model_U_net_2double_decoder can't improve the result quality.  I left only model_U_Net, and delete the variations of different U_Nets.

**Comment by KaijieMo1** [comment]
merge dev_kaijie - master

**Review by thomaskuestner — APPROVED** [review]
